api4/user.go

```diff
@@ -1568,3 +1570,29 @@ func registerTermsOfServiceAction(c *Context, w http.ResponseWriter, r *http.Req
 	c.LogAudit("TermsOfServiceId=" + termsOfServiceId + ", accepted=" + strconv.FormatBool(accepted))
 	ReturnStatusOK(w)
 }
+
+func getUserAnalytics(c *Context, w http.ResponseWriter, r *http.Request) {
+	// Types of data sets to return
+	// All data sets are defined as Datasets in mattermost-redux/src/constants/user_analytics
+	// Data sets are mirrored in the struct, UserAnalytics, in model/user.go
+	types := r.URL.Query().Get("types")
+
+	// The id of the current mattermost team
+	teamId := r.URL.Query().Get("teamId")
+
+	// The user id of the logged in user
+	uid := r.URL.Query().Get("uid")
+
+	// Call to service layer (/app/user.go)
+	rows, err := c.App.GetUserAnalytics(uid, teamId, types)
+
+	// If service or data layer errored, return
+	// Error will get logged in function that errored
+	if err != nil {
+		c.Err = err
+		return
+	}
+
+	// Return user analytics data to mattermost-redux as json
+	w.Write([]byte(rows.ToJson()))
```

Comment:
What do you mean by 'exceptional failure modes'? You mean like, there aren't any concerns about errors on the line w.Write([]byte(rows.ToJson()))?

Reply:
I mean literally that byte writing would usually be something that could throw an exception, perhaps because some of the elements of `rows` can't be properly serialized to JSON, say. The documentation for `ResponseWriter.Write(...)` doesn't mention any such possibility, so probably it's fine, but I was surprised not to see logic for handling possible exceptions.

Reply:
I think I did such a thorough job of type checking throughout, and then defined a function explicitly for converting that type into JSON, but I agree, surprised other functions don't implement an error catch on that statement.